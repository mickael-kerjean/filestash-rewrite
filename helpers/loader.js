import { get as getRelease } from "../pages/adminpage/model_release.js";

let version = null;

export async function loadScript(url) {
    const $script = document.createElement("script");
    $script.setAttribute("src", url);
    document.head.appendChild($script);
    return new Promise((done) => {
        $script.onload = done;
        $script.onerror = done;
    });
}

export async function CSS(baseURL, ...arrayOfFilenames) {
    const sheets = await Promise.all(arrayOfFilenames.map((filename) => loadSingleCSS(baseURL, filename)));
    return sheets.join("\n\n");
}

async function loadSingleCSS(baseURL, filename) {
    const res = await fetch(baseURL.replace(/(.*)\/[^\/]+$/, "$1/") + `${filename}?version=${version}`, {
        cache: "force-cache",
    });
    if (res.status !== 200) return `/* ERROR: ${res.status} */`;
    else if (!res.headers.get("Content-Type")?.startsWith("text/css")) return `/* ERROR: wrong type, got "${res.headers?.get("Content-Type")}"*/`;
    return await res.text();
}

export async function init() {
    const info = await getRelease().toPromise();
    version = info.version;
}
