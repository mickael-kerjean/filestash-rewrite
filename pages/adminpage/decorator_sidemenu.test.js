import { createElement } from "../../lib/skeleton/index.js";
import hoc from "./decorator_sidemenu.js";
import rxjs from "../../lib/rx.js";
import { CSS } from "../../helpers/loader.js";

import { get as getRelease } from "./model_release.js";
jest.mock("./model_release.js");
jest.mock("../../helpers/loader.js");

const version = "__version__";

describe("admin::middleware::sidemenu", () => {
    beforeEach(() => {
        jest.clearAllMocks();
        getRelease.mockImplementation(() => rxjs.of({ version }));
        CSS.mockImplementation(() => Promise.resolve(""));
    });
    afterEach(() => {
        global.location = global.window.location;
    });

    it("render a side menu", async () => {
        // given
        const ctrl = hoc((render) => render(createElement(`<div id="children"></div>`)));
        const render = jest.fn();

        // when
        await ctrl(render);

        // then
        expect(render).toBeCalledTimes(1);
        const $page = render.mock.calls[0][0];
        expect($page.querySelector("#children")).toBeTruthy();
        expect($page.querySelectorAll("a").length).toBe(5); // Backend, Settings, Logs, About, Quit
        expect($page.innerHTML).toContain(version);
    });

    it("snapshot", async () => {
        // given
        const ctrl = hoc((render) => render(createElement(`<div id="children"></div>`)));
        const render = jest.fn();
        global.location = { pathname: "/admin/settings" };

        // when
        await ctrl(render);

        // then
        expect(render.mock.calls[0][0]).toMatchSnapshot();
    });
});
