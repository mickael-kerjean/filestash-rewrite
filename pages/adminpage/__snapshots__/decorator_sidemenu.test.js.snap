// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin::middleware::sidemenu snapshot 1`] = `
<div
  class="component_page_admin"
>
  
                
  <style />
  
                
  <div
    class="component_menu_sidebar no-select"
  >
    
                    
    <a
      class="header"
      href="/"
    >
      
                        
      <svg
        class="arrow_left"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        
                            
        <path
          d="m 16,7.16 -4.58,4.59 4.58,4.59 -1.41,1.41 -6,-6 6,-6 z"
        />
        
                        
      </svg>
      
                        
      <span
        data-bind="logo"
      >
        <svg
          class="logo"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          
                     
          <path
            d="M330 202a81 79 0 00-162 0 81 79 0 000 158 81 79 0 000-158m81 79a81 79 0 1181 79H168"
            fill="none"
            stroke="currentColor"
            stroke-width="35px"
          />
          
                 
        </svg>
      </span>
      
                    
    </a>
    
                    
    <h2>
      Admin console
    </h2>
    
                    
    <ul>
      
                        
      <li>
        
                            
        <a
          data-link=""
          href="/admin/backend"
        >
          
                                Backend
                            
        </a>
        
                        
      </li>
      
                        
      <li>
        
                            
        <a
          class="active"
          data-link=""
        >
          
                                Settings
                            
        </a>
        
                        
      </li>
      
                        
      <li>
        
                            
        <a
          data-link=""
          href="/admin/logs"
        >
          
                                Logs
                            
        </a>
        
                        
      </li>
      
                        
      <li
        class="version"
      >
        
                            
        <a
          data-bind="version"
          data-link=""
          href="/admin/about"
        >
          __version__
        </a>
        
                        
      </li>
      
                    
    </ul>
    
                
  </div>
  
                
  <div
    class="page_container scroll-y"
    data-bind="admin"
  >
    <div
      id="children"
    />
  </div>
  
            
</div>
`;
